spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java

```diff
@@ -102,20 +107,25 @@
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Datastore datastore() {
-		DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
-				.setProjectId(this.projectId)
-				.setHeaderProvider(new UserAgentHeaderProvider(this.getClass()))
-				.setCredentials(this.credentials);
-		if (this.namespace != null) {
-			builder.setNamespace(this.namespace);
-		}
-
-		if (this.host != null) {
-			builder.setHost(this.host);
+	public DatastoreProvider datastoreSupplier(
+			ObjectProvider<DatastoreNamespaceProvider> namespaceProvider,
+			ObjectProvider<Datastore> datastoreProvider) {
+		if (datastoreProvider.getIfAvailable() != null) {
+			namespaceProvider.ifAvailable(unused -> {
+				throw new DatastoreDataException(
+						"A Datastore namespace provider and Datastore client were both configured. " +
+								"Only one can be configured.");
+			});
+			return datastoreProvider::getIfAvailable;
 		}
+		return getDatastoreProvider(
+				namespaceProvider.getIfAvailable() == null ? () -> this.namespace : namespaceProvider.getIfAvailable());
```

Comment:
I think you can simplify this line, if you change the order of the beans like this:

```java
	@Bean
	@ConditionalOnMissingBean({ Datastore.class, DatastoreNamespaceProvider.class })
	public Datastore datastore() {
		return getDatastore(this.namespace);
	}

	@Bean
	@ConditionalOnMissingBean
	public DatastoreProvider datastoreSupplier(
			ObjectProvider<DatastoreNamespaceProvider> namespaceProvider,
			ObjectProvider<Datastore> datastoreProvider) {
		if (datastoreProvider.getIfAvailable() != null) {
			namespaceProvider.ifAvailable(unused -> {
				throw new DatastoreDataException(
						"A Datastore namespace provider and Datastore client were both configured. " +
								"Only one can be configured.");
			});
			return datastoreProvider::getIfAvailable;
		}
		return getDatastoreProvider(namespaceProvider.getIfAvailable());
	}
```

There would be no way to end up here with `namespaceProvider` not existing.

Reply:
completed!